rl_studio/agents/autoparking/train_qlearn.py

```diff
@@ -0,0 +1,329 @@
+
+        # Checking state and actions
+        print_messages(
+            "In train_qlearn.py",
+            actions=self.actions,
+            # epsilon_decay=epsilon_decay,
+        )
+
+        ## -------------    START TRAINING --------------------
+        print(LETS_GO)
+        for episode in tqdm(
+            range(1, self.total_episodes + 1), ascii=True, unit="episodes"
+        ):
+            done = False
+            cumulated_reward = 0
+            step = 0
+            start_time_epoch = datetime.now()
+
+            observation = self.env.reset()
+            state = "".join(map(str, observation))
+
+            # ------- WHILE
+            while not done:
+
+                step += 1
+
+                # Pick an action based on the current state
+                action = qlearn.selectAction(state)
+
+                # Execute the action and get feedback
+                observation, reward, done, _ = self.env.step(action)
+                cumulated_reward += reward
+                next_state = "".join(map(str, observation))
+
+                # qlearning
+                qlearn.learn(state, action, reward, next_state)
+
+                ## important!!!
+                state = next_state
+
+                # render params
+                render_params(
+                    episode=episode,
+                    step=step,
+                    v=self.actions[action][0],
+                    w=self.actions[action][1],
+                    epsilon=self.epsilon,
+                    observation=observation,
+                    reward_in_step=reward,
+                    cumulated_reward=cumulated_reward,
+                    done=done,
+                )
+
+                # -------------------------------------- stats
+                try:
+                    self.states_counter[next_state] += 1
+                except KeyError:
+                    self.states_counter[next_state] = 1
+
+                self.steps_in_every_epoch[int(episode)] = step
+                self.states_reward[int(episode)] = cumulated_reward
+
+                # End epoch
+                if step > self.estimated_steps:
+                    done = True
+                    # self.min_reward = cumulated_reward
+                    print_messages(
+                        "ended training",
+                        epoch_time=datetime.now() - start_time_epoch,
+                        training_time=datetime.now() - start_time,
+                        episode=episode,
+                        episode_reward=cumulated_reward,
+                        steps=step,
+                        estimated_steps=self.estimated_steps,
+                        start_time=start_time,
+                        step_time=datetime.now()
+                        - timedelta(seconds=self.training_time),
+                    )
+                    if self.min_reward < cumulated_reward:
+                        self.min_reward = cumulated_reward
```

Comment:
Should it be named self.max_reward?

Reply:
min_reward is correct. When the training has finished because it has reached the limit of defined steps, we check its result and only those that exceed the threshold defined in min_reward will be recorded.